src/planner.cc

```diff
@@ -52,6 +52,37 @@ struct AxisTarget {
 };
 }  // end anonymous namespace
 
+// Given the desired target speed(or acceleration) of the defining axis and the
+// steps to be performed on all axes, determine if we need to scale down as to
+// not exceed the individual maximum speed or acceleration constraints on any
+// axis. Return the new defining axis limit based on the euclidean motion
+// requested.
+//
+// Example with speed: given i the axis with the highest (relative) out of
+// bounds speed, what's the speed of the defining_axis so that every speed
+// respects i's bounds? We just need to project each speed to the defining
+// axis speed, and pick the smallest.
+//
+// We need to work in mm, since steps is not a uniform unit of space across
+// the different axes.
+static float clamp_defining_axis_limit(const int *axes_steps,
+                                       const FloatAxisConfig &axes_limits_mm,
+                                       enum GCodeParserAxis defining_axis,
+                                       const FloatAxisConfig &steps_per_mm) {
+  float new_defining_axis_limit = axes_limits_mm[defining_axis];
+  for (const GCodeParserAxis i : AllAxes()) {
+    const float ratio =
+      std::fabs((1.0 * axes_steps[i] * steps_per_mm[defining_axis]) /
+                (axes_steps[defining_axis] * steps_per_mm[i]));
+
```

Comment:
so looking at the line above, we  can only be zero if the numerator is zero. Given that we can assume that the defining axis has `steps_per_mm` defined we could bail out early with a `continue`, so that below we don't have do do the awkward zero check:

```c++
for(...) {
  if (axes_steps[i] == 0) continue;
  const float ratio = ...;
  new_defining_axis_limit = std::min(new_defining_axis_limit, axes_limits_mm[i] / ratio);
}
```

Reply:
(solved)